src/Build/Graph/ProjectGraph.cs

```diff
@@ -479,28 +673,19 @@ private static IReadOnlyCollection<ProjectGraphNode> TopologicalSort(IReadOnlyCo
                     projectToEvaluate = projectsToEvaluate.Dequeue();
                     var task = new Task(() =>
                     {
-                        ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection, projectInstanceFactory);
-                        IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.ProjectInstance.GetItems(MSBuildConstants.ProjectReferenceItemName);
-                        foreach (var projectReferenceToParse in projectReferenceItems)
-                        {
-                            if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
-                            {
-                                throw new InvalidOperationException(string.Format(
-                                    CultureInfo.InvariantCulture,
-                                    ResourceUtilities.GetResourceString(
-                                        "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
-                                    projectReferenceToParse.EvaluatedInclude,
-                                    parsedProject.ProjectInstance.FullPath));
-                            }
+                        var parsedProject = CreateNewNode(projectToEvaluate, projectCollection, projectInstanceFactory, allParsedProjects);
 
-                            string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
-                            PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, projectToEvaluate.GlobalProperties);
-                            var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
-                            if (!tasksInProgress.ContainsKey(projectReferenceConfigurationMetadata))
+                        foreach (var referenceConfig in projectInterpretation.GetReferences(parsedProject.ProjectInstance))
+                        {
+                             /*todo: fix the following double check-then-act concurrency bug: one thread can pass the two checks, loose context,
+                             meanwhile another thread passes the same checks with the same data and inserts its reference. The initial thread regains context
+                             and duplicates the information, leading to wasted work
```

Comment:
Just wasted work, not a correctness problem? I can live with that for a while.

Reply:
As far as I can tell, not a correctness problem.